src/dense_layout.rs

```diff
@@ -0,0 +1,213 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2022
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+#![allow(clippy::too_many_arguments)]
+
+use ahash::RandomState;
+use hashbrown::{HashMap, HashSet};
+use indexmap::IndexSet;
+use ndarray::prelude::*;
+use numpy::PyReadonlyArray2;
+use numpy::ToPyArray;
+use rayon::prelude::*;
+
+use pyo3::prelude::*;
+use pyo3::wrap_pyfunction;
+use pyo3::Python;
+
+struct SubsetResult {
+    pub count: usize,
+    pub error: f64,
+    pub map: Vec<usize>,
+    pub subgraph: Vec<[usize; 2]>,
+}
+
+fn bfs_sort(adj_matrix: ArrayView2<f64>, start: usize) -> Vec<usize> {
+    let n = adj_matrix.shape()[0];
+    let mut next_level: IndexSet<usize, RandomState> =
+        IndexSet::with_hasher(RandomState::default());
+    let mut bfs_order = Vec::with_capacity(n);
+    let mut seen: HashSet<usize> = HashSet::with_capacity(n);
+    next_level.insert(start);
+    while !next_level.is_empty() {
+        let this_level = next_level;
+        next_level = IndexSet::with_hasher(RandomState::default());
+        let mut found: Vec<usize> = Vec::new();
+        for v in this_level {
+            if !seen.contains(&v) {
+                seen.insert(v);
+                found.push(v);
+                bfs_order.push(v);
+            }
+        }
+        if seen.len() == n {
+            return bfs_order;
+        }
+        for node in found {
+            for (idx, v) in adj_matrix.index_axis(Axis(0), node).iter().enumerate() {
+                if *v != 0. {
+                    next_level.insert(idx);
+                }
+            }
+            for (idx, v) in adj_matrix.index_axis(Axis(1), node).iter().enumerate() {
+                if *v != 0. {
+                    next_level.insert(idx);
+                }
+            }
+        }
+    }
+    bfs_order
+}
+
+/// Find the best subset in the coupling graph
+///
+/// This function will find the best densely connected subgraph in the
+/// coupling graph to run the circuit on. It factors in measurement error and
+/// cx error if specified.
+///
+/// Args:
+///
+///     num_qubits (int): The number of circuit qubits
+///     coupling_adjacency (numpy.ndarray): An adjacency matrix for the
+///         coupling graph.
+///     num_meas (int): The number of measurement operations in the circuit
+///     num_cx (int): The number of CXGates that are in the circuit
+///     use_error (bool): Set to True to use the error
+///     symmetric_coupling_map (bool): Is the coupling graph symmetric
```

Comment:
Yeah, that's what it means here. In the physical sense it means all the cx gates in the coupling map are bidirectional. I made this a standalone parameter here because the coupling graph can cache this (and there is a retworkx native function to compute it quickly). This just influences how we define the error rate/score of a layout, if this is true we divide the cx error rate in half because there will be 2 for each pair of nodes in the graph.